### ISSUE TITLE
Categorical data support.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#6503]
We started initial experimental support for categorical in xgboost 1.3.  The initial target is to make the one-hot encoding based tree split available for all xgboost components.  Here is a list of to-do items:

## One Hot
- Recognize categorical data in DMatrix.
  + [X] Accept feature types.
  + [X] DMatrix binary.
  + [X] Revert data handler. (#5893)
  + [x] Support for non-dataframe types (https://github.com/dmlc/xgboost/pull/7220)

- Remove min values.
  + [ ] Clarify trivial split (https://github.com/dmlc/xgboost/issues/2914).
  + [ ] Replace trivial split with infinity.

- Device DMatrix.
  + [x] Normal data. (#7042)
  + [x] Iterator. (#7043)

- GPU Sketching for categorical data.
  + [X] Count categories automatically.
  + [x] Weighted sketching. (#6508)
  + [x] Handle merging by removing duplicated entries after merge. (#7043)
  + [X] Ignore num bins in GPU pruning. (#6137)
  + [X] Handle compress index. (#6140)

- GPU Evaluate splits based on feature types.
  + [X] Unify split candidate structs. (https://github.com/dmlc/xgboost/pull/6037)
  + [X] Working implementation.
  + [ ] Handle monotone constraint.

- Tree model (RegTree).
  + [X] Define expand categorical.
  + [x] Tree model dump. (#7036)
  + [x] Feature scores. (#7041)
  + [x] Model as dataframe. (#7065)
  + [x] Split value histogram. (#7065)

- GPU Hist Updater.
  + [X] Evaluate split
  + [X] Apply split

- Prediction.
  + [x] Use predictor in dart (https://github.com/dmlc/xgboost/pull/6693).
  + [x] CPU Predictor. (#7001)
  + [X] GPU Predictor. (#6165)
  + [x] CPU leaf prediction. (#7001)
  + [x] GPU leaf prediction. (#7001)
  + [X] Missing value handling.
  + [x] Shap. (#7053)
  + [x] In-place prediction. (https://github.com/dmlc/xgboost/pull/7322)

- Model IO.
  + [X] Design JSON model schema.
  + [X] Implement it.
  + [X] Update schema.
  + [X] Set JSON to be the default format for pickle serialization.

- CPU Sketching.
  + [X] Merge sparse cut builder and dense cut builder. (#5880)
  + [x] Count categories. (https://github.com/dmlc/xgboost/pull/7221)
  + [x] Merge operation. (https://github.com/dmlc/xgboost/pull/7575)
  + [x] Check for invalid categorical values. (https://github.com/dmlc/xgboost/pull/7576)

- CPU Hist.
  + [x] Evaluate split (https://github.com/dmlc/xgboost/pull/7393)
  + [x] Update partition (https://github.com/dmlc/xgboost/pull/7695)

- Migrate the implementation of global approx to hist's codebase.
  + [x] Implement external memory support with `GHistIndexMatrix` used by `hist`.  (#7064, #7183) .
  + [x] Make evaluation function from hist reusable. (#7079) .
  + [x] Make histogram builder reusble. (https://github.com/dmlc/xgboost/pull/7152)
  + [x] Make part of the partitioner reusble.
  + [x] Modify sketching function to support hessian as weights. (#7081)

- Global approx (https://github.com/dmlc/xgboost/pull/7214)
  + [x] Gradient Index (https://github.com/dmlc/xgboost/pull/7371)
  + [x] Evaluator (https://github.com/dmlc/xgboost/pull/7393)
  + [x] Partitioner

- Local approx.
Unsupported.

- Sketch maker.
  + [X] Deprecate it.

- Exact
Unsupported.

- [x] Sklearn interfaces (#7011).

- Dask
  + [x] functional (#7043)
  + [x] sklearn (#7054)
  + [x] In-place prediction. (https://github.com/dmlc/xgboost/pull/7708)
  + [x] Dask dafaframe cat codes.

- Documents
  + [x] Brief introduction to native model support. (https://github.com/dmlc/xgboost/pull/7307)
  + [x] Simple demo. (https://github.com/dmlc/xgboost/pull/7213)

## Partitioning Based (LGB)
- General
  + [x] Let tree method be aware of whether it can be used (https://github.com/dmlc/xgboost/pull/7385)

- GPU Hist
  + [ ] Rewrite GPU evaluation into 2 passes evaluation style. (https://github.com/dmlc/xgboost/pull/7197)
  + [x] Evaluator (https://github.com/dmlc/xgboost/pull/7652)
  + [ ] Explore tunning parameters.

- Approx
  + [x] Evaluate split (https://github.com/dmlc/xgboost/pull/7393)
  + [x] Distributed (https://github.com/dmlc/xgboost/pull/7575)
  + [x] Tests (https://github.com/dmlc/xgboost/pull/7569)

- CPU Hist
  + [x] Evaluate split (https://github.com/dmlc/xgboost/pull/7393)

- Documents
  + [x] Demo (https://github.com/dmlc/xgboost/pull/7652)
  + [x] Tutorial (https://github.com/dmlc/xgboost/pull/7652)

Feature items can only be marked as completed if there's a corresponding (unit)test.  Please let me know if there are missing items or if you want to help to accelerate the progress. ;-)

**Comment by trivialfis** [comment]
As of posting, I have prototypes on new GPU evaluation function and refactored CPU hist/approx/local tree methods.  Dart support is on the way.

**Comment by mayer79** [comment]
Great idea! Will this be compatible with

- interaction constraints?
- monotonic constraints (probably not)?
- SHAP contribution calculations?

**Comment by hcho3** [comment]
> interaction constraints?

Yes, since we only care whether two features appear together or not.

> monotonic constraints 

No, since categorical values cannot be sorted in increasing order.

> SHAP contribution calculations

Probably, given that LightGBM also supports computing SHAP with categorical features. But we need to test it.

**Comment by trivialfis** [comment]
Status update:

Here are the few big items remaining for feature completeness:
- GPU evaluation function rewrite to have better performance.
- Add partition-based cat split for the rewritten evaluation function.
- Extract the partitioner from hist and implement categorical data support in it.
- Add categorical feature specific regularization parameters.

**Comment by trivialfis** [comment]
Closing, initial support is completed.  We will continue to add optimization and new features in the future.
